### PR TITLE
Implementation of migration function for provenance blockchain

### PR DESCRIPTION
**Pull request by iramiller**
## Description

This PR implements a custom migration function for existing v039 provenance blockchains to the current public chain.  The migration was tested against an export file from an existing chain.

The follow command used a 2GB export json file and processed successfully.
```
./build/provenanced -t migrate v1.0.0 ./local_path/export.json > migrate.json 
```

- Fixes an issue in the marker migration where untyped marker_type values would cause a panic
- Fixes an issue with metadata export that used `encoding/json` instead of Amino json
- Fixes a gogo proto compatibility issue with proto Timestamp
- Fixes migration of metadata groups to prevent null points
- Fixes an issue with GroupSpecification identifier being set to Record Group type in migration

closes: #78 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: internal/legacy/v040/migrate.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+		// the respective key.
+		appState[v040gentypes.ModuleName] = v040Codec.MustMarshalJSON(migrateGenutil(genutilGenState))
+	}
+
+	// Default capabilities initialization
+	capGenesis := captypes.DefaultGenesis()
+	appState[captypes.ModuleName] = v040Codec.MustMarshalJSON(capGenesis)
+
+	/*
+		TODO: default configuration for IBC
```
Comment (iramiller) [inline internal/legacy/v040/migrate.go:241]: this is a gaia config... it is not clear that we actually want this.  I filed an issue #102 for future follow up.

**Comment by iramiller** [comment]
this one needs 👀 from @dpederson-figure on the idea of using the `internal` folder for encapsulating our custom migration functions.

**Comment by dpederson-figure** [comment]
The `internal` package is a nice place to put things to reduce public API surface. So, if we don't want to expose the migration types/functions, it is a great place for them.

**Review by dpederson-figure — APPROVED** [review]
LGTM

**Pull request merged**